presto-native-sidecar-plugin/src/main/java/com/facebook/presto/sidecar/typemanager/NativeTypeManager.java

```diff
@@ -0,0 +1,298 @@
+import static com.facebook.presto.common.type.StandardTypes.VARCHAR;
+import static com.facebook.presto.common.type.VarcharType.createUnboundedVarcharType;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class NativeTypeManager
+        implements TypeManager
+{
+    private static final Logger log = Logger.get(NativeTypeManager.class);
+
+    private static final Set<String> NATIVE_ENGINE_SUPPORTED_TYPES =
+            ImmutableSet.of(
+                    BIGINT,
+                    REAL,
+                    VARBINARY,
+                    TIMESTAMP,
+                    TINYINT,
+                    BOOLEAN,
+                    DATE,
+                    INTEGER,
+                    DOUBLE,
+                    SMALLINT,
+                    HYPER_LOG_LOG,
+                    JSON,
+                    TIMESTAMP_WITH_TIME_ZONE,
+                    UUID,
+                    IPADDRESS,
+                    IPPREFIX,
+                    INTERVAL_DAY_TO_SECOND,
+                    INTERVAL_YEAR_TO_MONTH,
+                    VARCHAR,
+                    UNKNOWN);
+
+    private static final Set<String> NATIVE_ENGINE_SUPPORTED_PARAMETRIC_TYPES =
+            ImmutableSet.of(
+                    ARRAY,
+                    DECIMAL,
+                    MAP,
+                    ROW,
+                    FUNCTION);
+
+    private final TypeManager typeManager;
+    private final LoadingCache<ExactTypeSignature, Type> parametricTypeCache;
+    private final ConcurrentMap<TypeSignature, Type> types = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, ParametricType> parametricTypes = new ConcurrentHashMap<>();
+
+    @Inject
+    public NativeTypeManager(TypeManager typeManager)
+    {
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        parametricTypeCache = CacheBuilder.newBuilder()
+                .maximumSize(1000)
+                .build(CacheLoader.from(this::instantiateParametricType));
+        addAllTypes(
+                filterSupportedTypes(NATIVE_ENGINE_SUPPORTED_TYPES, typeManager.getTypes(), Type::getDisplayName),
+                filterSupportedTypes(
+                        NATIVE_ENGINE_SUPPORTED_PARAMETRIC_TYPES,
+                        new ArrayList<>(typeManager.getParametricTypes()),
+                        ParametricType::getName));
+    }
+
+    @Override
+    public Type getType(TypeSignature typeSignature)
+    {
+        // Todo: Fix this hack, native execution does not support parameterized char/varchar type signatures.
+        if (typeSignature.getBase().equals("char") || typeSignature.getBase().equals("varchar")) {
```

Comment:
nit: use the type name from the StandardTypes class
```suggestion
        if (typeSignature.getBase().equals(CHAR) || typeSignature.getBase().equals(VARCHAR)) {
```

Reply:
Done, thanks !